python/ray/data/_internal/execution/operators/actor_pool_submitter.py

```diff
@@ -0,0 +1,179 @@
+    ) -> Tuple[Block, BlockMetadata]:
+        # Coalesce all fn output blocks.
+        # TODO(Clark): Remove this coalescing once dynamic block splitting is supported
+        # for actors.
+        # yield from _map_task(fn, *blocks)
+        stats = BlockExecStats.builder()
+        builder = DelegatingBlockBuilder()
+        for block in fn(iter(blocks)):
+            builder.add_block(block)
+        block = builder.build()
+        block_metadata = BlockAccessor.for_block(block).get_metadata([], None)
+        block_metadata.exec_stats = stats.build()
+        return block, block_metadata
+
+
+class ActorPool:
+    """A pool of actors for map task execution.
+
+    This class is in charge of tracking the number of in-flight tasks per actor,
+    providing the least heavily loaded actor to the task submitter, and killing idle
+    actors when the task submitter is done submitting work to the pool.
+    """
+
+    def __init__(self):
+        # Number of tasks in flight per actor.
+        self._tasks_in_flight: Dict[ray.actor.ActorHandle, int] = {}
+        # Whether actors that become idle should be eagerly killed. This is False until
+        # the first call to kill_idle_actors().
+        self._should_kill_idle_actors = False
+
+    def add_actor(self, actor: ray.actor.ActorHandle):
+        """Adds an actor to the pool."""
+        self._tasks_in_flight[actor] = 0
+
+    def pick_actor(self) -> ray.actor.ActorHandle:
+        """Provides the least heavily loaded actor in the pool for task submission."""
+        actor = min(
+            self._tasks_in_flight.keys(), key=lambda actor: self._tasks_in_flight[actor]
+        )
+        self._tasks_in_flight[actor] += 1
+        return actor
+
+    def return_actor(self, actor: ray.actor.ActorHandle):
+        """Returns the provided actor to the pool."""
+        self._tasks_in_flight[actor] -= 1
+        if self._should_kill_idle_actors and self._tasks_in_flight[actor] == 0:
+            self._kill_actor(actor)
+
+    def has_actor(self, actor: ray.actor.ActorHandle) -> bool:
+        """Whether the provided actor is in this pool."""
+        return actor in self._tasks_in_flight
+
+    def get_tasks_in_flight(self, actor: ray.actor.ActorHandle) -> int:
+        """The number of tasks in flight for the provided actor."""
+        return self._tasks_in_flight[actor]
+
+    @property
+    def num_actors(self) -> int:
+        """The number of actors in the pool."""
+        return len(self._tasks_in_flight)
+
+    def kill_idle_actors(self):
+        """Kills all currently idle actors.
+
+        This is called once the task submitter is done submitting work to the pool.
+        """
+        idle_actors = [
+            actor
+            for actor, tasks_in_flight in self._tasks_in_flight.items()
+            if tasks_in_flight == 0
+        ]
+        for actor in idle_actors:
+            self._kill_actor(actor)
+
+    def kill_future_idle_actors(self):
```

Comment:
Can this be absorbed into `kill_idle_actors()`? One less interface is nicer if makes sense to combine.

Reply:
Sure, these two methods have pretty high cohesion and should probably just be a single method.